### ISSUE TITLE
Decode Jpeg on background thread

### ISSUE THREAD

**Issue by fschaus** [onevcat/Kingfisher#1084]
Hello,

When running setImage with Jpeg source files, I get a huge performance hit. Things get worst as it is all happening on the main thread. From what I understand UIImage needs to convert JPEG files to PNG before being able to use them.

I was wondering if (a) kf provides a way to decode Jpeg to PNG and if so (b) whether there is any way to run the decoding on a background thread?

I saw in the doc talks about using a cache serializer to force caching the images as Png, which I tried to combine with the prefetch Download manager but that didn't lead to any improvements.

I also use .backgroundDecode but that still leaves a lot of decoding happening on the main thread.

Profiling of the main thread as I scroll through tableview cells with images:
<img width="954" alt="screen shot 2019-01-05 at 4 17 01 pm" src="https://user-images.githubusercontent.com/36870504/50730589-a0b16600-1105-11e9-83b4-b5ced788f15f.png">

It would be great if you could point me to the right processor/ APIs to move the JPEG decoding on a background thread.

Thanks for the help!

**Comment by fschaus** [comment]
Thanks for the reply! Yes I am using 4.10.1. I got it to work - I had a lot of images (main feed) being decoded on the main thread at display time. Which explains why I was seeing lots of JPEG decoding traces in my main thread. 

I looked more carefully through the documentation and found the .backgroundDecode which works great. Thanks for that! The only challenge is that it is quite aggressive with the memory (as you outlined on the release not for v5.0.0) and can sometime crash the app. 

I'll try v5.0.0 and if it's indeed a bit less aggressive on the memory, I think it will be the perfect combination. 

The only issue I still have is doing background decoding of gifs - it seem to only decode the first image and not the sequence and thus shows a still image instead of a gif. 

Am I missing something or is gif decoding on background thread not supported? Not critical for my use-case since I have a gif to display only every so often (and so can afford to decode it on the main thread) but just curious if I am missing something.

Thanks again for a great lib!
